File: algo/src/test/kotlin/com/hartwig/actin/algo/evaluation/laboratory/HasLimitedBilirubinPercentageOfTotalTest.java
package com.hartwig.actin.algo.evaluation.laboratory;

import static com.hartwig.actin.algo.evaluation.EvaluationAssert.assertEvaluation;

import java.time.LocalDate;

import com.hartwig.actin.PatientRecord;
import com.hartwig.actin.TestDataFactory;
import com.hartwig.actin.algo.datamodel.EvaluationResult;
import com.hartwig.actin.clinical.datamodel.ImmutableLabValue;
import com.hartwig.actin.clinical.datamodel.LabValue;
import com.hartwig.actin.clinical.interpretation.LabMeasurement;

import org.junit.Test;

public class HasLimitedBilirubinPercentageOfTotalTest {

    @Test
    public void canEvaluate() {
        HasLimitedBilirubinPercentageOfTotal function = new HasLimitedBilirubinPercentageOfTotal(50D, LocalDate.of(2020, 3, 3));

        LabValue validBilirubin =
                LabTestFactory.forMeasurement(LabMeasurement.TOTAL_BILIRUBIN).date(LocalDate.of(2020, 4, 4)).value(10).build();
        PatientRecord valid = LabTestFactory.withLabValue(validBilirubin);

        ImmutableLabValue.Builder directBili = LabTestFactory.forMeasurement(LabMeasurement.DIRECT_BILIRUBIN);
        assertEvaluation(EvaluationResult.PASS, function.evaluate(valid, directBili.value(3D).build()));
        assertEvaluation(EvaluationResult.FAIL, function.evaluate(valid, directBili.value(8D).build()));

        // Cannot determine if no total bilirubin
        assertEvaluation(EvaluationResult.UNDETERMINED,
                function.evaluate(TestDataFactory.createMinimalTestPatientRecord(), directBili.build()));

        // Cannot determine in case of old bilirubin.
        LabValue invalidBilirubin =
                LabTestFactory.forMeasurement(LabMeasurement.TOTAL_BILIRUBIN).date(LocalDate.of(2019, 4, 4)).value(10).build();
        PatientRecord invalid = LabTestFactory.withLabValue(invalidBilirubin);
        assertEvaluation(EvaluationResult.UNDETERMINED, function.evaluate(invalid, directBili.value(3D).build()));
    }

    @Test(expected = IllegalStateException.class)
    public void crashOnWrongInputLabValue() {
        HasLimitedBilirubinPercentageOfTotal function = new HasLimitedBilirubinPercentageOfTotal(50D, LocalDate.of(2020, 3, 3));

        function.evaluate(TestDataFactory.createMinimalTestPatientRecord(), LabTestFactory.forMeasurement(LabMeasurement.ALBUMIN).build());
    }
}

File: algo/src/test/kotlin/com/hartwig/actin/algo/evaluation/laboratory/HasLimitedBilirubinPercentageOfTotalTest.kt
package com.hartwig.actin.algo.evaluation.laboratory

import com.hartwig.actin.algo.evaluation.EvaluationAssert.assertEvaluation
import com.hartwig.actin.datamodel.clinical.LabMeasurement
import com.hartwig.actin.datamodel.TestPatientFactory
import com.hartwig.actin.datamodel.algo.EvaluationResult
import com.hartwig.actin.datamodel.clinical.LabValue
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import java.time.LocalDate

class HasLimitedBilirubinPercentageOfTotalTest {

    @Test
    fun canEvaluate() {
        val function = HasLimitedBilirubinPercentageOfTotal(50.0, LocalDate.of(2020, 3, 3))
        val validBilirubin: LabValue =
            LabTestFactory.create(LabMeasurement.TOTAL_BILIRUBIN, 10.0, LocalDate.of(2020, 4, 4))
        val valid = LabTestFactory.withLabValue(validBilirubin)
        assertEvaluation(
            EvaluationResult.PASS,
            function.evaluate(valid, LabMeasurement.DIRECT_BILIRUBIN, LabTestFactory.create(LabMeasurement.DIRECT_BILIRUBIN, 3.0))
        )
        val actual = function.evaluate(valid, LabMeasurement.DIRECT_BILIRUBIN, LabTestFactory.create(LabMeasurement.DIRECT_BILIRUBIN, 8.0))
        assertEvaluation(EvaluationResult.FAIL, actual)
        assertThat(actual.recoverable).isTrue

        // Cannot determine if no total bilirubin
        assertEvaluation(
            EvaluationResult.UNDETERMINED,
            function.evaluate(
                TestPatientFactory.createMinimalTestWGSPatientRecord(),
                LabMeasurement.DIRECT_BILIRUBIN,
                LabTestFactory.create(LabMeasurement.DIRECT_BILIRUBIN)
            )
        )

        // Cannot determine in case of old bilirubin.
        val invalidBilirubin: LabValue = LabTestFactory.create(LabMeasurement.TOTAL_BILIRUBIN, 10.0, LocalDate.of(2019, 4, 4))
        val invalid = LabTestFactory.withLabValue(invalidBilirubin)
        assertEvaluation(
            EvaluationResult.UNDETERMINED,
            function.evaluate(invalid, LabMeasurement.DIRECT_BILIRUBIN, LabTestFactory.create(LabMeasurement.DIRECT_BILIRUBIN, 3.0))
        )
    }

    @Test(expected = IllegalStateException::class)
    fun crashOnWrongInputLabValue() {
        val function = HasLimitedBilirubinPercentageOfTotal(50.0, LocalDate.of(2020, 3, 3))
        function.evaluate(
            TestPatientFactory.createMinimalTestWGSPatientRecord(),
            LabMeasurement.ALBUMIN,
            LabTestFactory.create(LabMeasurement.ALBUMIN)
        )
    }
}